### PR TITLE
feat: add selector

### PR DESCRIPTION
**Pull request by aymeric-duchein**
This is a proposition regarding those issues: https://github.com/ngxs-labs/testing/issues/3 and https://github.com/ngxs-labs/testing/issues/4 .

If this looks good, I'll add the docs. For now you can find an exemple in the spec.ts file.

This allow you to mock every `selector`* using the `mockSelect`  function. It returns a `Subject` for each `selector` allowing you to control in wich order, when and what each `selector` should emit.
You'll still need to import the `NgxsModule.forRoot()` module in the Testbed (you'll most likely need it anyway if your component dispatch anything to the store 😉  ) . 

 *`Selectors` with parameters should be exported into constant beforehand as function comparison won't recognise them otherwise

### CODE REVIEW

File: src/tests/ngxs.setup.spec.ts [hunk 1]
```diff
@@ -136,3 +142,81 @@ describe('Full testing NGXS States with NgxsTestBed', () => {
...
+
+    let foodSelectorSubject: Subject<number>;
+    let animalsSelectorSubject: Subject<number>;
+    let fixture: ComponentFixture<HostComponent>;
+    let component: HostComponent;
+
+    beforeEach(() => {
+        TestBed.configureTestingModule({
+            declarations: [HostComponent],
+            imports: [CommonModule, NgxsModule.forRoot()]
```
Comment (splincode) [inline src/tests/ngxs.setup.spec.ts]: why `forRoot()` instead `forRoot([ZooState])`

**Comment by splincode** [comment]
What do you think about?

```ts
import { mockSelect } from '@ngxs-labs/testing/jest'; // sub package for jest

beforeEach(() => {
        TestBed.configureTestingModule({
            declarations: [HostComponent],
            imports: [CommonModule, NgxsModule.forRoot([ZooState])]
        }).compileComponents();

        foodSelector$ =mockSelect(ZooState.feed);
        animalsSelector$ =mockSelect(ZooState.animals);

        fixture = TestBed.createComponent(HostComponent);
        component = fixture.componentInstance;

        fixture.detectChanges();
    });
```

File: src/tests/ngxs.setup.spec.ts (hunk 1 above)
Comment (aymeric-duchein) [reply src/tests/ngxs.setup.spec.ts]: We don't need to import it as we are mocking the selectors. 
By doing so, we don't have to import every dependency of our states. 
With this you can test you component by testing wheter or not you are dispatching the correct actions or displaying the correct template value when an event occurs from a selector, an user input or something like that. 
It's my approach to unit testing, but it should still work for tests including the side effects triggered by the store, both can work together.

**Comment by aymeric-duchein** [comment]
> What do you think about?
> 
> ```ts
> import { mockSelect } from '@ngxs-labs/testing/jest'; // sub package for jest
> 
> beforeEach(() => {
>         TestBed.configureTestingModule({
>             declarations: [HostComponent],
>             imports: [CommonModule, NgxsModule.forRoot([ZooState])]
>         }).compileComponents();
> 
>         foodSelector$ =mockSelect(ZooState.feed);
>         animalsSelector$ =mockSelect(ZooState.animals);
> 
>         fixture = TestBed.createComponent(Hostponent);
>         component = fixture.componentInstance;
> 
>         fixture.detectChanges();
>     });
> `

This would be cleaner indeed, a jasmine version would be nice to have now that you mention subpackaging

File: src/tests/ngxs.setup.spec.ts (hunk 1 above)
Comment (splincode) [reply src/tests/ngxs.setup.spec.ts]: it’s just confusing if I test the condition itself

**Comment by splincode** [comment]
@aymeric-duchein any progress?

**Comment by aymeric-duchein** [comment]
Updated with the new Jest sub-package.

File: src/lib/ngxs.setup.ts [hunk 2]
```diff
@@ -136,4 +138,24 @@ export class NgxsTestBed {
 
         adapter.appendChild(document.body, root);
     }
+
+    static mockSelect<T>(selectorFn: (state: any) => T): Subject<T> {
```
Comment (splincode) [inline src/lib/ngxs.setup.ts]: why is it here in the main class?


Only need 

```ts
import { mockSelect } from '@ngxs-labs/testing/jest';
```

instead 

```ts
import { NgxsTestBed } from '@ngxs-labs/testing'
```

File: src/tests/ngxs.setup.spec.ts (hunk 1 above)
Comment (splincode) [inline src/tests/ngxs.setup.spec.ts]: Please add

```ts
NgxsModule.forRoot([ ZooState ])
```

**Review by splincode — CHANGES_REQUESTED** [review]

File: src/tests/ngxs.setup.spec.ts [hunk 3]
```diff
@@ -1,5 +1,11 @@
-import { Action, NgxsAfterBootstrap, NgxsOnInit, State, StateContext } from '@ngxs/store';
+import { Action, NgxsAfterBootstrap, NgxsModule, NgxsOnInit, Select, Selector, State, StateContext } from '@ngxs/store';
 import { NgxsTestBed } from '../lib/ngxs.setup';
+import { Component } from '@angular/core';
+import { Observable, Subject } from 'rxjs';
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+import { By } from '@angular/platform-browser';
+import { CommonModule } from '@angular/common';
+import { mockSelect } from '../jest/src/public_api';
```
Comment (splincode) [inline src/tests/ngxs.setup.spec.ts]: maybe `import { mockSelect } from '@ngxs-labs/testing/jest'`

Comment (aymeric-duchein) [reply src/tests/ngxs.setup.spec.ts]: This won't work as it is inside the lib (the same goes for the NgxsTestBed import at line 2 `import { NgxsTestBed } from '../lib/ngxs.setup';
`)
Do you want me to add path aliases?

Comment (splincode) [reply src/tests/ngxs.setup.spec.ts]: > Do you want me to add path aliases?

Yes

Comment (aymeric-duchein) [reply src/tests/ngxs.setup.spec.ts]: Done 👍

**Review by splincode — APPROVED** [review]

**Pull request merged**